src/algorithms/rsa.rs

```diff
@@ -34,97 +33,86 @@ pub fn rsa_encrypt<K: PublicKeyParts>(key: &K, m: &BigUint) -> Result<BigUint> {
+    let d = priv_key.d();
 
-    if priv_key.n().is_zero() {
+    if c >= n.as_ref() {
         return Err(Error::Decryption);
     }
 
     let mut ir = None;
 
+    let n_params = priv_key.n_params();
+    let bits = d.bits_precision();
+
     let c = if let Some(ref mut rng) = rng {
-        let (blinded, unblinder) = blind(rng, priv_key, c);
+        let (blinded, unblinder) = blind(rng, priv_key, c, n_params);
         ir = Some(unblinder);
-        Cow::Owned(blinded)
+        blinded.widen(bits)
     } else {
-        Cow::Borrowed(c)
+        c.widen(bits)
     };
 
-    let dp = priv_key.dp();
-    let dq = priv_key.dq();
-    let qinv = priv_key.qinv();
-    let crt_values = priv_key.crt_values();
-
-    let m = match (dp, dq, qinv, crt_values) {
-        (Some(dp), Some(dq), Some(qinv), Some(crt_values)) => {
-            // We have the precalculated values needed for the CRT.
-
-            let p = &priv_key.primes()[0];
-            let q = &priv_key.primes()[1];
-
-            let mut m = c.modpow(dp, p).into_bigint().unwrap();
-            let mut m2 = c.modpow(dq, q).into_bigint().unwrap();
-
-            m -= &m2;
-
-            let mut primes: Vec<_> = priv_key
-                .primes()
-                .iter()
-                .map(ToBigInt::to_bigint)
-                .map(Option::unwrap)
-                .collect();
-
-            while m.is_negative() {
-                m += &primes[0];
-            }
-            m *= qinv;
-            m %= &primes[0];
-            m *= &primes[1];
-            m += &m2;
-
-            let mut c = c.into_owned().into_bigint().unwrap();
-            for (i, value) in crt_values.iter().enumerate() {
-                let prime = &primes[2 + i];
-                m2 = c.modpow(&value.exp, prime);
-                m2 -= &m;
-                m2 *= &value.coeff;
-                m2 %= prime;
-                while m2.is_negative() {
-                    m2 += prime;
-                }
-                m2 *= &value.r;
-                m += &m2;
-            }
-
-            // clear tmp values
-            for prime in primes.iter_mut() {
-                prime.zeroize();
-            }
-            primes.clear();
-            c.zeroize();
-            m2.zeroize();
-
-            m.into_biguint().expect("failed to decrypt")
-        }
-        _ => c.modpow(priv_key.d(), priv_key.n()),
+    let has_precomputes = priv_key.dp().is_some();
+    let is_multiprime = priv_key.primes().len() > 2;
+
+    let m = if is_multiprime || !has_precomputes {
+        // c^d (mod n)
+        pow_mod_params(&c, d, n_params)
+    } else {
+        // We have the precalculated values needed for the CRT.
+
+        let dp = priv_key.dp().expect("precomputed");
+        let dq = priv_key.dq().expect("precomputed");
+        let qinv = priv_key.qinv().expect("precomputed");
+        let p_params = priv_key.p_params().expect("precomputed");
+        let q_params = priv_key.q_params().expect("precomputed");
```

Comment:
The previous pattern-based approach seemed a little more elegant than this

Reply:
fixed